recirq/quantum_chess/swap_updater.py

```diff
@@ -39,6 +39,37 @@ def _satisfies_adjacency(gate: cirq.Operation) -> bool:
     return q1.is_adjacent(q2)
 
 
+def _pairwise_shortest_distances(
+    qubits: Iterable[cirq.GridQubit]
+) -> Dict[Tuple[cirq.GridQubit, cirq.GridQubit], int]:
+    """Precomputes the shortest path length between each pair of qubits.
+
+    Returns:
+        dictionary mapping a pair of qubits to the length of the shortest path
+        between them (disconnected qubits will be absent).
+    """
+    # Do BFS starting from each qubit and collect the shortest path lengths as
+    # we go.
+    # For device graphs where all edges are the same (where each edge can be
+    # treated as unit length) repeated BFS is O(v**2) and is faster than
```

Comment:
Since this is exponential in the number of edges, should we consider aborting if the number of edges is above a sanity limit?  If you believe the hype, we are not _that_ far away from much larger qubit counts.

Reply:
each round of bfs visits each vertex and edge at most once so O(V+E), and we do bfs V times. That's O(V**2 + V E) ie quadratic in the number of edges and vertices, not exponential

Assuming with bigger devices qubits are still likely to be locally connected (like grids where the device graph is sparse and edges per vertex bounded by a constant) then E is O(V) and this function is O(V**2) overall. Not sure if that assumption will withstand the test of time, but at least on a graph of 'GridQubit' it seems like a safe assumption. Even a fully connected graph would have at most V**2 edges making repeated bfs cubic in the number of qubits O(V**3) not exponential.

We're precomputing O(V**2) vertex-to-vertex distance values so it is highly unlikely that we'll be able to do any better on sparse device graphs or grids :)